genesis/engine/solvers/rigid/array_class.py

```diff
@@ -4,14 +4,248 @@
 import taichi as ti
 
 import genesis as gs
+import numpy as np
 
 # we will use struct for DofsState and DofsInfo after Hugh adds array_struct feature to taichi
 DofsState = ti.template()
 DofsInfo = ti.template()
+GeomsState = ti.template()
+GeomsInfo = ti.template()
+GeomsInitAABB = ti.template()
+LinksState = ti.template()
+LinksInfo = ti.template()
+VertsInfo = ti.template()
+EdgesInfo = ti.template()
 
 
 @ti.data_oriented
 class RigidGlobalInfo:
-    def __init__(self, n_dofs: int, n_entities: int, n_geoms: int, f_batch: Callable):
+    def __init__(self, n_dofs: int, n_entities: int, n_geoms: int, _B: int, f_batch: Callable):
         self.n_awake_dofs = ti.field(dtype=gs.ti_int, shape=f_batch())
         self.awake_dofs = ti.field(dtype=gs.ti_int, shape=f_batch(n_dofs))
+
+        self.n_geoms = ti.field(dtype=gs.ti_int, shape=())
+        self.n_geoms[None] = n_geoms
+
+        self._B = ti.field(dtype=gs.ti_int, shape=())
+        self._B[None] = _B
+
+
+# =========================================== Collider ===========================================
+
+
+@ti.data_oriented
+class ColliderState:
+    """
+    Class to store the mutable collider data, all of which type is [ti.fields].
+    """
+
+    def __init__(self, solver, collider_info):
+        _B = solver._B
+        f_batch = solver._batch_shape
+        n_geoms = solver.n_geoms_
+        max_collision_pairs = solver._max_collision_pairs
+        use_hibernation = solver._static_rigid_sim_config.use_hibernation
+
+        ############## vertex connectivity ##############
+        self._init_verts_connectivity(solver)
+
+        ############## broad phase SAP ##############
+        # This buffer stores the AABBs along the search axis of all geoms
+        struct_sort_buffer = ti.types.struct(value=gs.ti_float, i_g=gs.ti_int, is_max=gs.ti_int)
+        self.sort_buffer = struct_sort_buffer.field(
+            shape=f_batch(2 * n_geoms),
+            layout=ti.Layout.SOA,
+        )
+
+        # This buffer stores indexes of active geoms during SAP search
+        if use_hibernation:
+            self.active_buffer_awake = ti.field(dtype=gs.ti_int, shape=f_batch(n_geoms))
+            self.active_buffer_hib = ti.field(dtype=gs.ti_int, shape=f_batch(n_geoms))
+        self.active_buffer = ti.field(dtype=gs.ti_int, shape=f_batch(n_geoms))
+
+        # Stores the validity of the collision pairs
+        self.collision_pair_validity = ti.field(dtype=gs.ti_int, shape=(n_geoms, n_geoms))
+        n_possible_pairs = self._init_collision_pair_validity(solver)
```

Comment:
Ideally, we want to separate the data from the logic. Ultimately, classes like ColliderState should be a pure dataclasses that only store data. The initialization logic should go into the Collider or RigidSolver class. Maybe I’ll push to your PR and show you what I mean.